File: lint-arwaky/src-python/taxonomy/file_path_vo.py
"""file_path_vo — File and directory path value objects."""

from pydantic import BaseModel, ConfigDict, field_validator, model_validator


class FilePath(BaseModel):
    """File path identifier."""

    model_config = ConfigDict(frozen=True)
    value: str

    @model_validator(mode="before")
    @classmethod
    def coerce_from_primitive(cls, data: object) -> object:
        if isinstance(data, str):
            return {"value": data}
        if isinstance(data, dict) and "value" in data:
            val = data["value"]
            if hasattr(val, "value") and isinstance(val.value, str):
                return {"value": val.value}
        if hasattr(data, "value") and isinstance(data.value, str):
            return {"value": data.value}
        return data

    @field_validator("value")
    @classmethod
    def check_non_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("File path cannot be empty")
        normalized = v.strip().replace("\\", "/").rstrip("/")
        if not normalized:
            if v.strip() == "/":
                return "/"
            return "."
        return normalized

    def __str__(self) -> str:
        return self.value

    def __hash__(self) -> int:
        return hash(self.value)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, FilePath):
            return self.value == other.value
        if isinstance(other, str):
            return self.value == other
        return NotImplemented

    @property
    def extension(self) -> str:
        """File extension without dot."""
        special_files = {
            "Makefile",
            "Dockerfile",
            "Dockerfile.dev",
            "Dockerfile.prod",
            ".bashrc",
            ".profile",
            ".zshrc",
            ".gitignore",
            ".dockerignore",
        }
        if self.value in special_files or self.value.startswith("."):
            return ""
        parts = self.value.rsplit(".", 1)
        return parts[-1] if len(parts) > 1 else ""

    def has_extension(self, ext: str) -> bool:
        """Check if path has given extension (without dot)."""
        return self.extension.lower() == ext.lower()


class DirectoryPath(BaseModel):
    """Directory path identifier."""

    model_config = ConfigDict(frozen=True)
    value: str

    @model_validator(mode="before")
    @classmethod
    def coerce_from_primitive(cls, data: object) -> object:
        if isinstance(data, str):
            return {"value": data}
        if isinstance(data, dict) and "value" in data:
            val = data["value"]
            if hasattr(val, "value") and isinstance(val.value, str):
                return {"value": val.value}
        if hasattr(data, "value") and isinstance(data.value, str):
            return {"value": data.value}
        return data

    @field_validator("value")
    @classmethod
    def check_non_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Directory path cannot be empty")
        return v.strip().replace("\\", "/").rstrip("/")

    def __str__(self) -> str:
        return self.value

    def __hash__(self) -> int:
        return hash(self.value)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, DirectoryPath):
            return self.value == other.value
        if isinstance(other, str):
            return self.value == other
        return NotImplemented

File: lint-arwaky/src-rust/taxonomy/file_path_vo.rs
use std::hash::{Hash, Hasher};

/// file_path_vo — File and directory path value objects.

/// File path identifier.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FilePath {
    value: String,
}

impl FilePath {
    /// Create a new FilePath from a string.
    ///
    /// # Errors
    /// Returns an error if the path is invalid (empty or only whitespace).
    pub fn new<S: Into<String>>(value: S) -> Result<Self, String> {
        let mut value = value.into();
        if value.trim().is_empty() {
            return Err("File path cannot be empty".to_string());
        }
        // Normalize: replace backslashes with forward slashes, and remove trailing slash.
        value = value.replace('\\', "/");
        // Remove trailing slash unless it's just "/"
        if value.ends_with('/') && value.len() > 1 {
            value.pop();
        }
        // If after normalization it's empty, then it was all slashes -> treat as root?
        if value.is_empty() {
            return Ok(FilePath { value: "/".to_string() });
        }
        Ok(FilePath { value })
    }

    /// File extension without dot.
    pub fn extension(&self) -> String {
        let special_files = [
            "Makefile",
            "Dockerfile",
            "Dockerfile.dev",
            "Dockerfile.prod",
            ".bashrc",
            ".profile",
            ".zshrc",
            ".gitignore",
            ".dockerignore",
        ];
        if special_files.contains(&self.value.as_str()) || self.value.starts_with('.') {
            return "".to_string();
        }
        match self.value.rsplitn(2, '.').nth(1) {
            Some(ext) => ext.to_string(),
            None => "".to_string(),
        }
    }

    /// Check if path has given extension (without dot).
    pub fn has_extension(&self, ext: &str) -> bool {
        self.extension().eq_ignore_ascii_case(ext)
    }
}

impl std::ops::Deref for FilePath {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.value
    }
}

impl std::fmt::Display for FilePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.value)
    }
}

impl Hash for FilePath {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}

/// Directory path identifier.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DirectoryPath {
    value: String,
}

impl DirectoryPath {
    /// Create a new DirectoryPath from a string.
    ///
    /// # Errors
    /// Returns an error if the path is invalid (empty or only whitespace).
    pub fn new<S: Into<String>>(value: S) -> Result<Self, String> {
        let mut value = value.into();
        if value.trim().is_empty() {
            return Err("Directory path cannot be empty".to_string());
        }
        // Normalize: replace backslashes with forward slashes, and remove trailing slash.
        value = value.replace('\\', "/");
        // Remove trailing slash unless it's just "/"
        if value.ends_with('/') && value.len() > 1 {
            value.pop();
        }
        Ok(DirectoryPath { value })
    }
}

impl std::ops::Deref for DirectoryPath {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.value
    }
}

impl std::fmt::Display for DirectoryPath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.value)
    }
}

impl Hash for DirectoryPath {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_file_path_new() {
        let fp = FilePath::new("test.txt").unwrap();
        assert_eq!(fp.value, "test.txt");
        assert_eq!(fp.extension(), "txt");
        assert!(fp.has_extension("txt"));
        assert!(!fp.has_extension("md"));

        // Test normalization
        let fp = FilePath::new("path\\to\\file.txt").unwrap();
        assert_eq!(fp.value, "path/to/file.txt");

        let fp = FilePath::new("path/to/file/").unwrap();
        assert_eq!(fp.value, "path/to/file");

        let fp = FilePath::new("/").unwrap();
        assert_eq!(fp.value, "/");

        let fp = FilePath::new("///").unwrap();
        assert_eq!(fp.value, "/");
    }

    #[test]
    fn test_file_path_invalid() {
        assert!(FilePath::new("").is_err());
        assert!(FilePath::new("   ").is_err());
    }

    #[test]
    fn test_directory_path_new() {
        let dp = DirectoryPath::new("test/dir").unwrap();
        assert_eq!(dp.value, "test/dir");

        let dp = DirectoryPath::new("test/dir/").unwrap();
        assert_eq!(dp.value, "test/dir");

        let dp = DirectoryPath::new("/").unwrap();
        assert_eq!(dp.value, "/");
    }

    #[test]
    fn test_directory_path_invalid() {
        assert!(DirectoryPath::new("").is_err());
        assert!(DirectoryPath::new("   ").is_err());
    }
}
